src/main/java/net/mcreator/workspace/references/ReferencesFinder.java

```diff
@@ -0,0 +1,309 @@
+	 * @return Whether the provided value or any value extracted from valid fields/methods on the {@code value} object
+	 * passes the provided condition.
+	 */
+	@SuppressWarnings("unchecked") private static <T> boolean checkValue(@Nullable Object value, AccessibleObject field,
+			Class<T> clazz, Predicate<AccessibleObject> validIf, BiPredicate<AccessibleObject, T> condition) {
+		if (value == null)
+			return false;
+
+		if (clazz.isInstance(value)) { // value of specified type
+			return (isCustomObject(value) || validIf == null || validIf.test(field)) && (condition == null
+					|| condition.test(field, (T) value));
+		} else if (clazz.isArray()) { // array of values
+			int length = Array.getLength(value);
+			for (int i = 0; i < length; i++) {
+				if (checkValue(Array.get(value, i), field, clazz, validIf, condition))
+					return true;
+			}
+		} else if (value instanceof Iterable<?> list) { // list of values
+			return listHasMatches(list, field, clazz, validIf, condition);
+		} else if (value instanceof Map<?, ?> map) { // map with values
+			return listHasMatches(map.keySet(), field, clazz, validIf, condition) || listHasMatches(map.values(), field,
+					clazz, validIf, condition);
+		} else if (isCustomObject(value)) { // value of unknown type
+			return anyValueMatches(value, clazz, validIf, condition);
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks if any value (or its components) on the list acquired from the passed field/method meets given condition.
+	 *
+	 * @param list      The extracted list of values that should be checked.
+	 * @param field     The field/method the {@code value} was acquired.
+	 * @param clazz     The class of values to be checked.
+	 * @param validIf   The predicate used to check if the field/method is considered valid.
+	 * @param condition The predicate defining the condition that the acquired values should pass.
+	 * @param <T>       The type of values to be checked.
+	 * @return Whether the provided value or any value extracted from valid fields/methods on the {@code value} object
+	 * passes the provided condition.
+	 */
+	private static <T> boolean listHasMatches(Iterable<?> list, AccessibleObject field, Class<T> clazz,
+			Predicate<AccessibleObject> validIf, BiPredicate<AccessibleObject, T> condition) {
+		for (Object obj : list) {
+			if (checkValue(obj, field, clazz, validIf, condition))
+				return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Checks if class of the passed value is from this module and not related to the technical part of the application.
+	 * Scanning values that do not pass this condition will most probably lead to a {@link StackOverflowError}.
+	 * <br>NOTE: If needed values are instances of a class not contained in this module, they will still be checked.
+	 *
+	 * @param value The value that should be checked.
+	 * @return Whether it is safe to scan the {@code value} object deeper.
+	 */
+	private static boolean isCustomObject(Object value) {

```

Comment:
One last change idea. Thoughts on `ReferenceProvider` interface/annotation (which one?) to allow scanning custom data structures (like loot table pools or trade entries) declared outside MCreator's module?

Reply:
How are trade entries outside MCreator's module (which is currently an unnamed module)?

Reply:
~Declared outside but functioning similar to e.g. trade entries.~
So unnamed module covers all packages without `module-info.class`?

Reply:
Yes.

However, I like the idea. I think the best would be the interface as it is passed down and would thus require fewer code changes. Do you have any estimate of how many files would need to change?

The only downside I see it this is another source of bugs for things that one forgot to add interface to

Reply:
This should not be needed for our classes because we don't really design subclasses of things like said loot table pools.

Reply:
So that'd give 1 added and one changed file.

Reply:
I would omit module check and instead use interface for all cases if we went this route (so either module only).

All plugins are already in an unnamed module too currently so the current system is plugin-compatible already.

But I agree interface approach (but in all cases), is clearer.

How many files would need changed this this case? I guess all entries, Procedure, MappableElement, quite some right?

Reply:
Loot table/villager trade pools/entries, effect entries, basically everything we define with `JEntriesList`s so far.